tfhe/src/core_crypto/entities/compressed_modulus_switched_glwe_ciphertext.rs

```diff
@@ -0,0 +1,261 @@
+            .iter()
+            .map(|a| modulus_switch(*a, log_modulus))
+            .collect();
+
+        let packed_integers = PackedIntegers::pack(&modulus_switched, log_modulus);
+
+        Self {
+            packed_integers,
+            glwe_dimension,
+            polynomial_size,
+            bodies_count,
+            uncompressed_ciphertext_modulus,
+        }
+    }
+
+    /// Converts back a compressed ciphertext to its initial modulus
+    /// The noise added during the compression stays in the output
+    /// The output must got through a PBS to reduce the noise
+    pub fn extract(&self) -> GlweCiphertextOwned<Scalar> {
+        let log_modulus = self.packed_integers.log_modulus.0;
+
+        let number_bits_to_unpack =
+            (self.glwe_dimension.0 * self.polynomial_size.0 + self.bodies_count) * log_modulus;
+
+        let len: usize = number_bits_to_unpack.div_ceil(Scalar::BITS);
+
+        assert_eq!(
+            self.packed_integers.packed_coeffs.len(), len,
+            "Mismatch between actual(={}) and maximal(={len}) CompressedModulusSwitchedGlweCiphertext packed_coeffs size",
+            self.packed_integers.packed_coeffs.len(),
+        );
+
+        let container = self
+            .packed_integers
+            .unpack()
+            // Scaling
+            .map(|a| a << (Scalar::BITS - log_modulus))
+            .chain(std::iter::repeat(Scalar::ZERO).take(self.polynomial_size.0 - self.bodies_count))
+            .collect();
+
+        GlweCiphertextOwned::from_container(
+            container,
+            self.polynomial_size,
+            self.uncompressed_ciphertext_modulus,
+        )
+    }
+}
+
+impl<Scalar: UnsignedInteger> ParameterSetConformant
+    for CompressedModulusSwitchedGlweCiphertext<Scalar>
+{
+    type ParameterSet = GlweCiphertextParameters<Scalar>;
+
+    fn is_conformant(&self, lwe_ct_parameters: &GlweCiphertextParameters<Scalar>) -> bool {
+        let log_modulus = self.packed_integers.log_modulus.0;
+
+        let number_bits_to_unpack =
+            (self.glwe_dimension.0 * self.polynomial_size.0 + self.bodies_count) * log_modulus;
+
+        let len = number_bits_to_unpack.div_ceil(Scalar::BITS);
+
+        self.packed_integers.packed_coeffs.len() == len
+            && self.glwe_dimension == lwe_ct_parameters.glwe_dim
+            && self.polynomial_size == lwe_ct_parameters.polynomial_size
+            && lwe_ct_parameters.ct_modulus.is_power_of_two()
+            && self.uncompressed_ciphertext_modulus == lwe_ct_parameters.ct_modulus
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::core_crypto::prelude::test::TestResources;
+
+    #[test]
+    fn glwe_ms_compression_() {
+        glwe_ms_compression::<u32>(1, GlweDimension(1), PolynomialSize(512), 0);
+        glwe_ms_compression::<u32>(10, GlweDimension(1), PolynomialSize(512), 1);
+        glwe_ms_compression::<u32>(11, GlweDimension(1), PolynomialSize(512), 511);
```

Comment:
create a file in the algorithms/test module similar to the modulus compression, try to have actual tests with encryption and decryption, hard code the test parameters to the shortint parameters if needed

Reply:
can be done in a backlog issue, we have integration tests on top https://github.com/zama-ai/tfhe-rs-internal/issues/618